openmdao/solvers/ln_direct.py

```diff
@@ -42,23 +91,120 @@ def solve(self, rhs_mat, system, mode):
+            if self.options['jacobian_method'] == 'MVP':
+
+                self.mode = mode
+
+                n_edge = len(rhs)
+                ident = np.eye(n_edge)
+
+                partials = np.empty((n_edge, n_edge))
+
+                for i in range(n_edge):
+                    partials[:, i] = self.mult(ident[:, i])
+
+            # Assemble the Jacobian
+            else:
+
+                # Must clear the jacobian if we switch modes
+                if self.mode != mode:
+                    self.setup(system)
+                self.mode = mode
+                sys_name = system.name + '.'
+                conn = system.connections
+
+                partials = self.jacobian
+                u_vec = system.unknowns
+
+                for sub in system.components(recurse=True, include_self=True):
+
+                    jac = sub._jacobian_cache
+
+                    # This method won't work on components where apply_linear
+                    # is overridden.
+                    if jac is None:
+                        msg = "The 'assemble' jacobian_method is not supported when " + \
+                             "'apply_linear' is used on a component (%s)." % sub.pathname
+                        raise RuntimeError(msg)
+
+                    sub_u = sub.unknowns
+                    sub_name = sub.pathname
+                    icache = self.icache
+                    sub_to_prom_name = sub._probdata.to_prom_name
+                    sys_to_prom_name = system._sysdata.to_prom_name
+
+                    for key in jac:
+                        o_var, i_var = key
+                        key2 = (sub_name, key)
+
+                        # Derivs wrt states, but states live in u_vec
+                        if i_var in sub.states:
+                            sub_p = sub_u
+                        else:
+                            sub_p = sub.params
+
+                        # Skip anything not relevant
+                        if o_var not in sub_u or i_var not in sub_p:
+                            continue
+
+                        # We cache the location of each variable in our jacobian
+                        if key2 not in icache:
+
+                            o_var_abs = '.'.join((sub_name, o_var))
+                            i_var_abs = '.'.join((sub_name, i_var))
+                            i_var_pro = sub_to_prom_name[i_var_abs]
+                            o_var_pro = sub_to_prom_name[o_var_abs]
+
+                            # States are fine ...
+                            if i_var in sub.states:
+                                pass
+
+                            #... but inputs need to find their source.
+                            elif i_var_pro not in u_vec:
+                                if i_var_pro in conn:
+                                    i_var_src = conn[i_var_pro][0]
+                                else:
+                                    i_var_src = conn[i_var_abs][0]
```

Comment:
i_var_abs will always be in conn, so don't bother checking if i_var_pro is in there and just use conn[i_var_abs][0]